internal/server/middleware.go

```diff
@@ -132,29 +131,44 @@ func validateOrgMatchesRequest(req *http.Request, tx data.GormTxn, accessKeyOrg
 	}
 }
 
-func withDBTxn(ctx context.Context, db *gorm.DB, fn func(tx *gorm.DB)) {
-	err := db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
-		fn(tx)
-		return nil
-	})
-	// TODO: https://github.com/infrahq/infra/issues/2697
+func withDBTxn(ctx context.Context, db *data.DB, fn func(tx *data.Transaction) error) error {
```

Comment:
This function implements roughly the same logic as `gorm.DB.Transaction`, but lets us call our transaction Commit and Rollback instead.

In the near future I expect us to remove this function, or at least integrate it into the route wrapper, so that we can resolve the issues linked in TODOs from the caller.